File: frontend/test/features/tenant_facility/domain/tenant_similarity_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:hosspi_hms/features/tenant_facility/domain/entities/tenant_facility_setup.dart';
import 'package:hosspi_hms/features/tenant_facility/domain/entities/tenant_similarity.dart';

void main() {
  group('checkTenantDuplicates', () {
    const List<TenantProfile> existing = <TenantProfile>[
      TenantProfile(
        id: 'TEN0001',
        name: 'DemoCare General Hospital',
        slug: 'democare-general-hospital',
        isActive: true,
        contactName: 'Jane Doe',
        contactEmail: 'jane@example.com',
        contactPhone: '+256700000000',
        currency: 'UGX',
        standardConsultationFee: '50000',
      ),
      TenantProfile(
        id: 'TEN0002',
        name: 'Acme Health',
        slug: 'acme-health',
        isActive: false,
      ),
    ];

    test('detects exact name and slug conflicts', () {
      final TenantDuplicateCheckResult result = checkTenantDuplicates(
        name: 'DemoCare General Hospital',
        slug: 'democare-general-hospital',
        existing: existing,
      );

      expect(result.exactNameConflict, isTrue);
      expect(result.exactSlugConflict, isTrue);
      expect(result.similarMatches, isNotEmpty);
      expect(result.overridableMatches, isEmpty);
    });

    test('detects similar tenant names above threshold', () {
      final TenantDuplicateCheckResult result = checkTenantDuplicates(
        name: 'Democare General Hospitl',
        slug: 'new-tenant',
        existing: existing,
      );

      expect(result.exactSlugConflict, isFalse);
      expect(result.nonExactSimilarMatches, isNotEmpty);
      expect(
        result.nonExactSimilarMatches.first.nameScore,
        greaterThanOrEqualTo(80),
      );
    });

    test('scores contact and config fields in the breakdown', () {
      final TenantDuplicateCheckResult result = checkTenantDuplicates(
        name: 'Democare General Hospitl',
        slug: 'new-tenant',
        contactName: 'Jane Doe',
        contactEmail: 'jane@example.com',
        contactPhone: '256700000000',
        currency: 'UGX',
        standardConsultationFee: '50000',
        existing: existing,
      );

      final TenantSimilarityMatch match = result.overridableMatches.first;
      expect(match.fieldComparisons, isNotEmpty);
      expect(
        match.fieldComparisons.any(
          (TenantFieldComparison comparison) =>
              comparison.field == 'contact_email' &&
              comparison.status == TenantFieldComparisonStatus.match,
        ),
        isTrue,
      );
    });

    test('ignores excluded tenant when editing', () {
      final TenantDuplicateCheckResult result = checkTenantDuplicates(
        name: 'DemoCare General Hospital',
        slug: 'democare-general-hospital',
        existing: existing,
        excludeTenantId: 'TEN0001',
      );

      expect(result.hasExactConflict, isFalse);
      expect(result.similarMatches, isEmpty);
    });

    test('ignores excluded tenant by mutation resource uuid', () {
      final List<TenantProfile> withUuid = <TenantProfile>[
        TenantProfile(
          id: 'TEN0001',
          name: 'DemoCare General Hospital',
          slug: 'democare-general-hospital',
          resourceUuid: 'uuid-tenant-1',
          contactName: 'Jane Doe',
          contactEmail: 'jane@example.com',
          contactPhone: '+256700000000',
          currency: 'UGX',
          standardConsultationFee: '50000',
        ),
      ];

      final TenantDuplicateCheckResult result = checkTenantDuplicates(
        name: 'DemoCare General Hospital',
        slug: 'democare-general-hospital',
        existing: withUuid,
        excludeTenantId: 'uuid-tenant-1',
      );

      expect(result.hasExactConflict, isFalse);
      expect(result.similarMatches, isEmpty);
    });
  });

  group('nameSimilarityScore', () {
    test('returns 100 for identical normalized names', () {
      expect(nameSimilarityScore('acme health', 'acme health'), 100);
    });
  });

  group('compositeTenantSimilarityScore', () {
    test('weights identity fields more than configuration fields', () {
      final int identityHeavy = compositeTenantSimilarityScore(
        nameScore: 100,
        slugScore: 100,
        currencyScore: 0,
        feeScore: 0,
      );
      final int configHeavy = compositeTenantSimilarityScore(
        nameScore: 0,
        slugScore: 0,
        currencyScore: 100,
        feeScore: 100,
      );
      expect(identityHeavy, greaterThan(configHeavy));
    });
  });
}

File: backend/src/tests/lib/tenant/tenant-similarity.test.js
const {
  checkTenantDuplicates,
  compositeSimilarityScore,
  normalizeEmail,
  normalizePhone,
  normalizeSlug,
  normalizeText
} = require('@lib/tenant/tenant-similarity');

describe('tenant-similarity', () => {
  const existing = [
    {
      id: 'tenant-1',
      human_friendly_id: 'TEN0001',
      name: 'DemoCare General Hospital',
      slug: 'democare-general-hospital',
      is_active: true,
      extension_json: {
        currency: 'UGX',
        contact: {
          name: 'Jane Doe',
          email: 'jane@example.com',
          phone: '+256 700 000 000'
        },
        billing: {
          standard_consultation_fee: '50000'
        }
      }
    }
  ];

  it('normalizes identity and contact values', () => {
    expect(normalizeText('  Demo-Care!! ')).toBe('democare');
    expect(normalizeSlug('Demo Care!!')).toBe('demo-care');
    expect(normalizeEmail(' Jane@Example.COM ')).toBe('jane@example.com');
    expect(normalizePhone('+256 700-000-000')).toBe('256700000000');
  });

  it('weights identity fields more heavily than configuration fields', () => {
    const identityHeavy = compositeSimilarityScore({
      nameScore: 100,
      slugScore: 100,
      currencyScore: 0,
      feeScore: 0
    });
    const configHeavy = compositeSimilarityScore({
      nameScore: 0,
      slugScore: 0,
      currencyScore: 100,
      feeScore: 100
    });
    expect(identityHeavy).toBeGreaterThan(configHeavy);
  });

  it('detects multi-field similar tenants with field comparisons', () => {
    const result = checkTenantDuplicates({
      name: 'Democare General Hospitl',
      slug: 'new-slug',
      contactName: 'Jane Doe',
      contactEmail: 'jane@example.com',
      contactPhone: '256700000000',
      currency: 'UGX',
      standardConsultationFee: 50000,
      existing
    });

    expect(result.exactSlugConflict).toBe(false);
    expect(result.nonExactSimilarMatches.length).toBeGreaterThan(0);
    expect(result.nonExactSimilarMatches[0].score).toBeGreaterThanOrEqual(70);
    expect(result.nonExactSimilarMatches[0].nameScore).toBeGreaterThanOrEqual(80);
    expect(result.nonExactSimilarMatches[0].field_comparisons).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ field: 'name' }),
        expect.objectContaining({ field: 'contact_email', status: 'MATCH' }),
        expect.objectContaining({ field: 'contact_phone', status: 'MATCH' })
      ])
    );
  });

  it('marks exact slug conflicts as non-overridable', () => {
    const result = checkTenantDuplicates({
      name: 'Completely Different',
      slug: 'democare-general-hospital',
      existing
    });

    expect(result.exactSlugConflict).toBe(true);
    expect(result.overridableMatches).toHaveLength(0);
    expect(result.similarMatches[0].exactSlugConflict).toBe(true);
  });
});
